### PR TITLE
Hardening lifecycle-state-store, name-store, and oci-hooks

### LINKED ISSUE #3351: ` name "foo" is already used by ID ""`
Looks like the name store may accidentally get stuck in an improper state
```console
$ sudo nerdctl run -it --rm --name=foo alpine
FATA[0000] name "foo" is already used by ID ""

$ sudo nerdctl rm -f foo
ERRO[0000] 1 errors:
no such container: foo

$ sudo nerdctl run -it --rm --name=foo alpine 
FATA[0000] name "foo" is already used by ID ""
```

Probably a recent regression in the main branch.

Workaround: `sudo rm /var/lib/nerdctl/1935db59/names/default/foo`

### LINKED ISSUE #3357: When run fails on (bridge) networking setup, it breaks the namestore, leading to duplicate names
### Description

Yet another variant / breakout of #3352 

### Steps to reproduce the issue

You need to edit onCteateRuntime so that it panics.

```
nerdctl rm -f $(nerdctl ps -aq)
nerdctl run --net bridge -d --name foo debian sleep Inf
systemctl --user stop containerd
systemctl --user start containerd
sleep 1
echo "restart foo"
nerdctl start foo
sleep 1
echo "run bar"
nerdctl run -d --name bar debian sleep Inf
nerdctl run -d --name bar debian sleep Inf
echo "result in duplication"
nerdctl ps -a
```

### Describe the results you received and expected

na

### What version of nerdctl are you using?

main

### Are you using a variant of nerdctl? (e.g., Rancher Desktop)

None

### Host information

_No response_

### PR DESCRIPTION
**Pull request by apostasie**
**First commit**

Although #3350 got closed as a fluke, the error message seen in the logs clearly shows that the lifecycle state file could end-up messed-up, triggering a failure.

This does two things:
- write atomically, so that we never end-up with half-written / hosed JSON in the file
- tolerate load errors: this file is not critical and purely informative and should soft fail with a warning instead

This will make the LifecycleState more robust and resistant to host shutdowns.

Note: we should consider a review of all our filesystem based stores and consider moving to atomic writing for them.

**Second commit**

Fix #3351 by making us resistant to the condition and allowing a container to acquire a "broken" name.

This obviously should never happen, but it does ^.

**Third commit**

Fix #3357 - does not fix the underlying issue, which is possibly upstream - but effectively mitigate it.
Since onPostStop is (wrongly?) called after an error in onCreateRuntime - while the container will NOT be deleted - we inhibit the normal cleanup flow in that case.

### CODE REVIEW

**Comment by apostasie** [comment]
Failure is the usual  IPFS/compose.

File: pkg/ocihook/state/state.go [hunk 1]
```diff
@@ -71,18 +74,26 @@ func (lf *LifecycleState) Load() error {
 	} else {
 		err = json.Unmarshal(data, lf)
 		if err != nil {
+			backout := filepath.Join(lf.stateDir, lifecycleFile, ".bak")
```
Comment (AkihiroSuda) [inline pkg/ocihook/state/state.go:77]: Maybe
```suggestion
			backout := filepath.Join(lf.stateDir, lifecycleFile + ".bak")
```
Also, where is this backout created?

Comment (apostasie) [reply pkg/ocihook/state/state.go:77]: Looks like I dropped a stash while rewriting commits.
Amending now.

Comment (apostasie) [reply pkg/ocihook/state/state.go]: Apologies for this.
Fixed.

**Comment by apostasie** [comment]
Grumble grumble.
Interactive rebase...
Should be ready in a second.

**Comment by apostasie** [comment]
@AkihiroSuda CI is green. Mistake addressed.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**